### PR TITLE
Add Hana wallet support

### PR DESCRIPTION
**Pull request by tzechuen**
Hi, I'm a developer for Hana wallet, currently available on the Chrome web store. 

We would like to add our wallet to the list of supported wallets for Astar portal. 

Please let me know if you have any questions.

Thanks,
Ting

### CODE REVIEW

**Review by sirius651 — APPROVED** [review]

**Comment by impelcrypto** [comment]
Hi @tzechuen 
Thank you for creating this PR. I've installed HanaWallet V2.6.1 and tried on localhost, but it looks like both EVM and native wallets are unable to connect to the portal. Have you tested connect wallets to the portal?

File: src/config/wallets.ts
Comment (impelcrypto) [inline src/config/wallets.ts:201]: `Provider` looks like `hanaWallet.ethereum` (`window.ethereum` connects to MetaMask).
image

**Comment by tzechuen** [comment]
@impelcrypto apologies, I've made some fixes and added Astar support, but it's not merged into production yet. Let me check with the rest of the team to see when they're planning to push another release and get back to you

File: src/config/wallets.ts
Comment (impelcrypto) [inline src/config/wallets.ts:25]: Please add `HanaNative` and `HanaEvm` here

**Review by impelcrypto — CHANGES_REQUESTED** [review]

**Comment by impelcrypto** [comment]
Please pull the latest main branch.

**Comment by gluneau** [comment]
I've tested the EVM portion of this PR.

Native token transfer
ASTR: https://astar.subscan.io/tx/0xc43c2eb2f11b5a5525437630f49c3912598d9c7d50e03d7397e52c4a947bf1fd

ERC20 token transfer
USDC: https://astar.subscan.io/tx/0x035fabb591ea00de9ba1bd9a6bee10c94a26a1d0669911e46691cbf4462c406d

XC20(XCM assets) token transfer
GLMR: https://astar.subscan.io/tx/0xd05a603a9c95cc0cbffb8d9b8af039fca4ab68f06dd5ef4e3887bfaec9a0043a

XCM transfer
https://astar.subscan.io/tx/0x5021b59d90480fb3da09b1aec51e9ceddf55dbcaefed3a2f5a15422717f8de72

Looks like the Native substrate portion needs more work before I can validate.

**Comment by impelcrypto** [comment]
Hi @gluneau , I meant what device did you use for testing? Chrome(computer) or Mobile?

**Comment by impelcrypto** [comment]
Hi @tzechuen Does this PR ready for reviewing?

**Comment by tzechuen** [comment]
Hi @impelcrypto sorry for late reply, we've been busy with implementing other features at the moment. I've just had a look at this again and it looks good to go -- however the extension in the Chrome store does not include the code that supports Astar yet and I assume we'll need to get that published before you merge this PR. Let me get back to you on this asap

**Comment by impelcrypto** [comment]
Hi @togamamora @Kahonnohak 
It would be great if you (and other ambassadors) can test Hana Wallet(Native and EVM) at https://astar-portal-hana.vercel.app/ 🙏

https://chrome.google.com/webstore/detail/hana-wallet/jfdlamikmbghhapbgfoogdffldioobgl

**Comment by impelcrypto** [comment]
I've updated HanaWallet extension to the latest version (2.9.2). However, it looks like `ASTAR` is not in the selectable chain options here.

image

image

image

**Comment by impelcrypto** [comment]
@tzechuen Sending tokens worked, but somehow the UI didn't show the loading animation like screenshot below. Do you know what might be the issue? Perhaps you can take a look [here](https://github.com/Hana-Technology/astar-apps/blob/add-hana-wallet/src/v2/services/implementations/PolkadotWalletService.ts#L180). 

image

**Comment by togamamora** [comment]
@impelcrypto 
Windows 11, Chrome.

I am able to see Hana on the options.
image

**Comment by togamamora** [comment]
OK. Same here, I cannot see it.

image

**Comment by Kahonnohak** [comment]
I can see the wallets both EVM and Native but when connecting to it I get this error
Screenshot 2023-04-21 at 11 19 41

**Comment by impelcrypto** [comment]
@Kahonnohak I think you saw this error when you reload the page.

**Comment by Kahonnohak** [comment]
Okay I managed to get it work and yes the loading is the issue, I tried to stake and the function is okay just did not get loading so it was not clear if the transaction went through or not

**Comment by impelcrypto** [comment]
@tzechuen 
Any updates for [this](https://github.com/AstarNetwork/astar-apps/pull/563#issuecomment-1517352791) comment?

**Comment by tzechuen** [comment]
@impelcrypto apologies we're taking a while, but yes my colleague has been able to reproduce the issue and we have the fixes planned. We have a lot going on at the moment but hopefully we will be able to sort this out in the next week or so

**Comment by impelcrypto** [comment]
@tzechuen 
I think you should take a look at [this](https://github.com/Hana-Technology/astar-apps/blob/add-hana-wallet/src/v2/services/implementations/PolkadotWalletService.ts#L179) line as well. Usually, substrate wallets should has 'signature' property in the `listener.data.response`

**Comment by tzechuen** [comment]
hey @impelcrypto sorry it took a while, but I've completed the fixes from my end and passed this on to my team for further testing so it shouldn't be too long. While we're at it, is this `listener.data.response` thing documented anywhere by Polkadot JS?

It seems to me these are internal relay messages they use to pass data between the extension and the background services. We also have a similar implementation, just that we didn't conform to the same standards Polkadot JS extension was using, hence the issue of this loading spinner not coming up on the site.

I ended up having to broadcast these messages although they aren't actually used by ourselves internally. Would love if you could point me to the documentation regarding this so I can understand if this was a required implementation for Polkadot JS wallets. Thanks!

**Comment by impelcrypto** [comment]
Hi @tzechuen I don't think there is a documentation about this but other substrate wallets(such as talisman, subwallet...) are using the same relay messages.

Maybe you take a look at this line from Polkadot.js portal.
https://github.com/polkadot-js/extension/blob/297b2af14c68574b24bb8fdeda2208c473eccf43/packages/extension/src/page.ts#L10-L22

**Comment by tzechuen** [comment]
hey @impelcrypto my team has pushed up the latest version of Hana Wallet to the Chrome Web Store which includes the fixes for the reported issues. Please have a go and let me know if you have any questions!

**Review by impelcrypto — APPROVED** [review]
LGTM

**Pull request merged**
